CANTroller2/src/qpid.h

```diff
@@ -447,12 +436,153 @@ float QPID::GetOutputSum() { return outputSum; }
 float QPID::GetOutputMin() { return outMin; }  // Soren
 float QPID::GetOutputMax() { return outMax; }  // Soren
 float QPID::GetOutputRange() { return outMax - outMin; }  // Soren
+float QPID::GetCenter() { return center; }  // Soren
 uint8_t QPID::GetMode() { return static_cast<uint8_t>(mode); }
 uint8_t QPID::GetDirection() { return static_cast<uint8_t>(action); }
 uint8_t QPID::GetPmode() { return static_cast<uint8_t>(pmode); }
 uint8_t QPID::GetDmode() { return static_cast<uint8_t>(dmode); }
 uint8_t QPID::GetAwMode() { return static_cast<uint8_t>(iawmode); }
 uint8_t QPID::GetCentMode() { return static_cast<uint8_t>(centmode); }  // Soren
-float QPID::GetCenter() { return center; }  // Soren
 
+class IdleControl {  // Soren - To allow creative control of PID targets in case your engine idle problems need that.
+  public:
+    enum class idlemodes : uint32_t { direct, control, minimize, num_idlemodes };  // direct: disable idle management.  control: soft landing to idle rpm.  minimize: attempt to minimize idle to edge of instability
+  protected:
+    enum targetstates : uint32_t { driving, droptohigh, droptolow, idling, stallpoint };
+    // String modenames[3] = { "direct", "cntrol", "minimz" };
+    // String statenames[4] = { "drivng", "tohigh", "tolow", "tostal" };
+    targetstates runstate, laststate;
+    idlemodes idlemode;
+    float targetlast_rpm, idle_rpm, idlehigh_rpm, idlehot_rpm, idlecold_rpm, stallpoint_rpm, dynamic_rpm, temphot_f, tempcold_f;
+    float margin_rpm = 10; float idle_absmax_rpm = 1000.0;  // High limit of idle speed adjustability
+    float* target_rpm; float* measraw_rpm; float* measfilt_rpm; float* coolant_f;
+    bool we_just_changed_states = true;
+    uint32_t settletime_us;
+    Timer settleTimer;
+  public:
+    IdleControl (float* target_rpm_arg, float* measraw_rpm_arg, float* measfilt_rpm_arg, float* coolant_f_arg,  // Variable references: idle target, rpm raw, rpm filt, Engine temp
+      float idlehigh, float idlehot, float idlecold,  // Values for: high-idle rpm (will not stall), hot idle nominal rpm, cold idle nominal rpm 
+      float tempcold, float temphot,  // Values for: engine operational temp cold (min) and temp hot (max) in degrees-f
+      uint32_t settletime_us = 2000000,  // Period over which the idle will be lowered from high-idle to final idle
+      idlemodes idlemode = idlemodes::control) {  // Configure idle control to just soft land or also attempt to minimize idle
+        target_rpm = target_rpm_arg;

```

Comment:
I think it looks better this way, and if we ever try to send null/messed up pointers it would break and let us know in testing... still might want to consider checking for it, eg 
```
if(target_rpm_arg != nullptr) {
    target_rpm = target_rpm_arg;
} else {
    // Handle error or throw exception
}
```
But doing that over and over is redundant. I dunno the best way to go about it, there's probably something though, I'll do some homework.